File: python/147. Insertion Sort List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        p = dummy
        while head:
            tmp = head
            head = head.next
            p = dummy
            while p.next and p.next.val <tmp.val:
                p = p.next
            k = p.next
            p.next = tmp
            tmp.next = k
        return dummy.next

File: c++/147. Insertion Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode *p = head, *new_head = new ListNode(0), *q, *t;
        while (p) {
            t = p;
            p = p->next;
            q = new_head;
            while(q->next && q->next->val < t->val)
                q = q->next;
            t->next = q->next;
            q->next = t;
        }
        
        head = new_head->next;
        delete(new_head);
        return head;
    }
};